@here/harp-mapview/lib/geometry/SolidLineMesh.ts

```diff
@@ -0,0 +1,451 @@
+            maxRSq,
+            center.distanceToSquared(vertex.fromBufferAttribute(pos, indices[i])),
+            center.distanceToSquared(vertex.fromBufferAttribute(pos, indices[i + VERTEX_STRIDE]))
+        );
+    }
+    sphere.radius = Math.sqrt(maxRSq);
+    return sphere;
+}
+
+/**
+ * Finds the intersection of a ray with a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param vExtrusion Line extrusion vector.
+ * @param normal Extrusion plane normal.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the extruded line intersection to the ray origin.
+ */
+function intersectExtrudedLine(
+    ray: THREE.Ray,
+    line: THREE.Line3,
+    vExtrusion: THREE.Vector3,
+    normal: THREE.Vector3,
+    hWidth: number
+): number {
+    const matrix = new THREE.Matrix4().makeBasis(
+        line.delta(new THREE.Vector3()).normalize(),
+        vExtrusion,
+        normal
+    );
+    const obb = new OrientedBox3(
+        line.getCenter(tmpV4),
+        matrix,
+        new THREE.Vector3(line.distance() / 2, hWidth, hWidth)
+    );
+    if (obb.contains(ray.origin)) {
+        return 0;
+    }
+    return obb.intersectsRay(ray) ?? Infinity;
+}
+
+/**
+ * Finds the intersection of a ray with the closest end cap of a extruded line.
+ * @param ray Intersection ray in object local space.
+ * @param line The centerline.
+ * @param hWidth Extrusion half width.
+ * @returns Distance of the end cap intersection to the ray origin.
+ */
+function intersectClosestEndCap(ray: THREE.Ray, line: THREE.Line3, hWidth: number): number {
+    const sphere = new THREE.Sphere(line.start, hWidth);
+    const startCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    sphere.center.copy(line.end);
+    const endCapT = sphere.containsPoint(ray.origin)
+        ? 0
+        : ray.intersectSphere(sphere, tmpV4)
+        ? tmpV4.sub(ray.origin).length()
+        : Infinity;
+    return Math.min(startCapT, endCapT);
+}
+
+/**
+ * Intersects line
+ * @param ray Intersection ray in object local space.
+ * @param line The line to intersect.
+ * @param vExtrusion Line extrusion vector.
+ * @param hWidth The line's extrusion half width.
+ * @param hWidthSq The line's extrusion half width squared.
+ * @param plane The extrusion plane.
+ * @param interPlane The intersection of the ray with the extrusion plane.
+ * @param interLine The ray intersetion with the extruded line.
+ * @returns true if ray intersects the extruded line, false otherwise.
+ */
+function intersectLine(
+    ray: THREE.Ray,
+    line: THREE.Line3,
+    vExtrusion: THREE.Vector3,
+    hWidth: number,
+    hWidthSq: number,
+    plane: THREE.Plane,
+    interPlane: THREE.Vector3,
+    interLine: THREE.Vector3
+): boolean {
+    if (interPlane.equals(tmpRay.origin) && tmpRay.direction.dot(plane.normal) === 0) {
```

Comment:
Using temp variables are an arguments here is quite fragile - where do you set it? - I don't see. Moreover we are dealing with floating point numbers, so dot product may be very sensitive to precision issues.

Reply:
> Using temp variables are an arguments here is quite fragile - where do you set it? - I don't see. 

done, it's actually the ray passed as parameter.

>Moreover we are dealing with floating point numbers, so dot product may be very sensitive to precision issues.

If the plane intersection is equal to the ray origin it can be only for two reasons:
- The ray is not coplanar but the origin is actually on the plane.
- The ray is coplanar and ray.intersectPlane() that I called before actually returned the ray's origin as the plane intersection. See here the actual THREE.Ray.distanceToPlane code which is called from ray.intersectPlane(), it only sets the distance to 0 (and therefore the intersection is at ray Origin) if the dot product is exactly 0:

`
distanceToPlane: function ( plane ) {

		var denominator = plane.normal.dot( this.direction );

		if ( denominator === 0 ) {

			// line is coplanar, return origin
			if ( plane.distanceToPoint( this.origin ) === 0 ) {

				return 0;

			}
`
In practice this means that we will almost always consider the picking ray as not coplanar, but that's fine. We don't need to be super precise here. What's the problem if you try to pick a extruded line with a ray coplanar to the extrusion plane (I mean, who tries to do this in the first place) and we fail to detect it's coplanar and miss the intersection? You can try to pick again.